### PR TITLE
[DispatchCreation] Limit attention broadcast fusion

### LINKED ISSUE #24051: iree-compile crashes on iree_linalg_ext.attention with broadcasting
`iree-compile` segfaults when compiling an SDPA op with broadcasted batch and head dims, where Q head dim size differs from V head dim size (d_qk=128, d_v=64).

Input IR:

<details>
<summary>input.mlir</summary>

```mlir
module @module {

  func.func private @sdpa_fprop(
      %arg0: !torch.vtensor<[8,2,64,128],bf16>,
      %arg1: !torch.vtensor<[8,2,16,128],bf16>,
      %arg2: !torch.vtensor<[8,2,16,64],bf16>)
      -> !torch.vtensor<[8,2,64,64],bf16> {
    %none_mask = torch.constant.none
    %dropout = torch.constant.float 0.000000e+00
    %is_causal = torch.constant.bool true
    %scale = torch.constant.float 8.838835e-02
    %enable_gqa = torch.constant.bool false
    %0 = torch.aten.scaled_dot_product_attention %arg0, %arg1, %arg2,
        %none_mask, %dropout, %is_causal, %scale, %enable_gqa :
        !torch.vtensor<[8,2,64,128],bf16>, !torch.vtensor<[8,2,16,128],bf16>,
        !torch.vtensor<[8,2,16,64],bf16>, !torch.none, !torch.float, !torch.bool,
        !torch.float, !torch.bool -> !torch.vtensor<[8,2,64,64],bf16>
    return %0 : !torch.vtensor<[8,2,64,64],bf16>
  }

  func.func @main(%o_3_: !torch.tensor<[2,8,64,64],bf16>, %k_1: !torch.vtensor<[1,1,16,128],bf16>, %q_0: !torch.vtensor<[1,1,64,128],bf16>, %v_2: !torch.vtensor<[1,1,16,64],bf16>) attributes {torch.assume_strict_symbolic_shapes} {

    %permute_IN_0_val_0_sdpa_fprop_i0 = torch.constant.int 0
    %permute_IN_0_val_1_sdpa_fprop_i0 = torch.constant.int 1
    %permute_IN_0_val_2_sdpa_fprop_i0 = torch.constant.int 2
    %permute_IN_0_val_3_sdpa_fprop_i0 = torch.constant.int 3
    %permute_IN_0_sdpa_fprop_i0 = torch.prim.ListConstruct %permute_IN_0_val_0_sdpa_fprop_i0, %permute_IN_0_val_1_sdpa_fprop_i0, %permute_IN_0_val_2_sdpa_fprop_i0, %permute_IN_0_val_3_sdpa_fprop_i0 : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int>

    %q_0_sdpa_fprop_i0_perm_unexpanded = torch.aten.permute %q_0, %permute_IN_0_sdpa_fprop_i0 : !torch.vtensor<[1,1,64,128],bf16>, !torch.list<int> -> !torch.vtensor<[1,1,64,128],bf16>
      %expand_size_permute_IN_0_val_0_sdpa_fprop_i0 = torch.constant.int 8
    %expand_size_permute_IN_0_val_1_sdpa_fprop_i0 = torch.constant.int 2
    %expand_size_permute_IN_0_val_2_sdpa_fprop_i0 = torch.constant.int 64
    %expand_size_permute_IN_0_val_3_sdpa_fprop_i0 = torch.constant.int 128
    %expand_size_permute_IN_0_sdpa_fprop_i0 = torch.prim.ListConstruct %expand_size_permute_IN_0_val_0_sdpa_fprop_i0, %expand_size_permute_IN_0_val_1_sdpa_fprop_i0, %expand_size_permute_IN_0_val_2_sdpa_fprop_i0, %expand_size_permute_IN_0_val_3_sdpa_fprop_i0 : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int>
%expand_implicit_permute_IN_0_sdpa_fprop_i0 = torch.constant.bool false
    %q_0_sdpa_fprop_i0_perm = torch.aten.expand %q_0_sdpa_fprop_i0_perm_unexpanded, %expand_size_permute_IN_0_sdpa_fprop_i0, %expand_implicit_permute_IN_0_sdpa_fprop_i0 : !torch.vtensor<[1,1,64,128],bf16>, !torch.list<int>, !torch.bool -> !torch.vtensor<[8,2,64,128],bf16>

    %permute_IN_1_val_0_sdpa_fprop_i1 = torch.constant.int 0
    %permute_IN_1_val_1_sdpa_fprop_i1 = torch.constant.int 1
    %permute_IN_1_val_2_sdpa_fprop_i1 = torch.constant.int 2
    %permute_IN_1_val_3_sdpa_fprop_i1 = torch.constant.int 3
    %permute_IN_1_sdpa_fprop_i1 = torch.prim.ListConstruct %permute_IN_1_val_0_sdpa_fprop_i1, %permute_IN_1_val_1_sdpa_fprop_i1, %permute_IN_1_val_2_sdpa_fprop_i1, %permute_IN_1_val_3_sdpa_fprop_i1 : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int>

    %k_1_sdpa_fprop_i1_perm_unexpanded = torch.aten.permute %k_1, %permute_IN_1_sdpa_fprop_i1 : !torch.vtensor<[1,1,16,128],bf16>, !torch.list<int> -> !torch.vtensor<[1,1,16,128],bf16>
      %expand_size_permute_IN_1_val_0_sdpa_fprop_i1 = torch.constant.int 8
    %expand_size_permute_IN_1_val_1_sdpa_fprop_i1 = torch.constant.int 2
    %expand_size_permute_IN_1_val_2_sdpa_fprop_i1 = torch.constant.int 16
    %expand_size_permute_IN_1_val_3_sdpa_fprop_i1 = torch.constant.int 128
    %expand_size_permute_IN_1_sdpa_fprop_i1 = torch.prim.ListConstruct %expand_size_permute_IN_1_val_0_sdpa_fprop_i1, %expand_size_permute_IN_1_val_1_sdpa_fprop_i1, %expand_size_permute_IN_1_val_2_sdpa_fprop_i1, %expand_size_permute_IN_1_val_3_sdpa_fprop_i1 : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int>
%expand_implicit_permute_IN_1_sdpa_fprop_i1 = torch.constant.bool false
    %k_1_sdpa_fprop_i1_perm = torch.aten.expand %k_1_sdpa_fprop_i1_perm_unexpanded, %expand_size_permute_IN_1_sdpa_fprop_i1, %expand_implicit_permute_IN_1_sdpa_fprop_i1 : !torch.vtensor<[1,1,16,128],bf16>, !torch.list<int>, !torch.bool -> !torch.vtensor<[8,2,16,128],bf16>

    %permute_IN_2_val_0_sdpa_fprop_i2 = torch.constant.int 0
    %permute_IN_2_val_1_sdpa_fprop_i2 = torch.constant.int 1
    %permute_IN_2_val_2_sdpa_fprop_i2 = torch.constant.int 2
    %permute_IN_2_val_3_sdpa_fprop_i2 = torch.constant.int 3
    %permute_IN_2_sdpa_fprop_i2 = torch.prim.ListConstruct %permute_IN_2_val_0_sdpa_fprop_i2, %permute_IN_2_val_1_sdpa_fprop_i2, %permute_IN_2_val_2_sdpa_fprop_i2, %permute_IN_2_val_3_sdpa_fprop_i2 : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int>

    %v_2_sdpa_fprop_i2_perm_unexpanded = torch.aten.permute %v_2, %permute_IN_2_sdpa_fprop_i2 : !torch.vtensor<[1,1,16,64],bf16>, !torch.list<int> -> !torch.vtensor<[1,1,16,64],bf16>
      %expand_size_permute_IN_2_val_0_sdpa_fprop_i2 = torch.constant.int 8
    %expand_size_permute_IN_2_val_1_sdpa_fprop_i2 = torch.constant.int 2
    %expand_size_permute_IN_2_val_2_sdpa_fprop_i2 = torch.constant.int 16
    %expand_size_permute_IN_2_val_3_sdpa_fprop_i2 = torch.constant.int 64
    %expand_size_permute_IN_2_sdpa_fprop_i2 = torch.prim.ListConstruct %expand_size_permute_IN_2_val_0_sdpa_fprop_i2, %expand_size_permute_IN_2_val_1_sdpa_fprop_i2, %expand_size_permute_IN_2_val_2_sdpa_fprop_i2, %expand_size_permute_IN_2_val_3_sdpa_fprop_i2 : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int>
%expand_implicit_permute_IN_2_sdpa_fprop_i2 = torch.constant.bool false
    %v_2_sdpa_fprop_i2_perm = torch.aten.expand %v_2_sdpa_fprop_i2_perm_unexpanded, %expand_size_permute_IN_2_sdpa_fprop_i2, %expand_implicit_permute_IN_2_sdpa_fprop_i2 : !torch.vtensor<[1,1,16,64],bf16>, !torch.list<int>, !torch.bool -> !torch.vtensor<[8,2,16,64],bf16>

    %o_3_sdpa_fprop_perm = func.call @sdpa_fprop(%q_0_sdpa_fprop_i0_perm, %k_1_sdpa_fprop_i1_perm, %v_2_sdpa_fprop_i2_perm) : (!torch.vtensor<[8,2,64,128],bf16>, !torch.vtensor<[8,2,16,128],bf16>, !torch.vtensor<[8,2,16,64],bf16>) -> !torch.vtensor<[8,2,64,64],bf16>
    %permute_OUT_0_val_0_sdpa_fprop = torch.constant.int 1
    %permute_OUT_0_val_1_sdpa_fprop = torch.constant.int 0
    %permute_OUT_0_val_2_sdpa_fprop = torch.constant.int 2
    %permute_OUT_0_val_3_sdpa_fprop = torch.constant.int 3
    %permute_OUT_0_sdpa_fprop = torch.prim.ListConstruct %permute_OUT_0_val_0_sdpa_fprop, %permute_OUT_0_val_1_sdpa_fprop, %permute_OUT_0_val_2_sdpa_fprop, %permute_OUT_0_val_3_sdpa_fprop : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int>

    %o_3 = torch.aten.permute %o_3_sdpa_fprop_perm, %permute_OUT_0_sdpa_fprop : !torch.vtensor<[8,2,64,64],bf16>, !torch.list<int> -> !torch.vtensor<[2,8,64,64],bf16>

    torch.overwrite.tensor.contents %o_3 overwrites %o_3_ : !torch.vtensor<[2,8,64,64],bf16>, !torch.tensor<[2,8,64,64],bf16>

    return
  }
}
```

</details>

Compile command:
```
iree-compile input.mlir --iree-hal-target-backends=rocm --iree-rocm-target=gfx942 --iree-input-type=torch -o /dev/null
```

Stack trace (from a nightly install so some debug info is missing):
```
Please report issues to https://github.com/iree-org/iree/issues and include the crash backtrace.
 #0 0x00007fec96d22308 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) Signals.cpp:0:0
 #1 0x00007fec96d1f4ce llvm::sys::RunSignalHandlers() Signals.cpp:0:0
 #2 0x00007fec96d23206 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x00007fec91442520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007fec9cf6403d mlir::dispatchIndexOpFoldResult(mlir::OpFoldResult, llvm::SmallVectorImpl<mlir::Value>&, llvm::SmallVectorImpl<long>&) StaticValueUtils.cpp:0:0
 #5 0x00007fec98ba50e0 mlir::iree_compiler::(anonymous namespace)::TileAndDistributeToWorkgroupsUsingForallOpPass::runOnOperation() TileDispatchUsingForall.cpp:0:0
 #6 0x00007fec9742cc61 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) Pass.cpp:0:0
 #7 0x00007fec9742d979 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) Pass.cpp:0:0
 #8 0x00007fec97431d0a llvm::LogicalResult llvm::function_ref<llvm::LogicalResult (mlir::OpPassManager&, mlir::Operation*)>::callback_fn<mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_0>(long, mlir::OpPassManager&, mlir::Operation*) Pass.cpp:0:0
 #9 0x00007fec99454f04 mlir::iree_compiler::(anonymous namespace)::LLVMGPULowerExecutableTargetPass::runOnOperation() LLVMGPULowerExecutableTarget.cpp:0:0
#10 0x00007fec9742cc61 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) Pass.cpp:0:0
#11 0x00007fec9742d979 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) Pass.cpp:0:0
#12 0x00007fec97431d0a llvm::LogicalResult llvm::function_ref<llvm::LogicalResult (mlir::OpPassManager&, mlir::Operation*)>::callback_fn<mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_0>(long, mlir::OpPassManager&, mlir::Operation*) Pass.cpp:0:0
#13 0x00007fec96eced69 mlir::iree_compiler::detail::OpPipelineAdaptorPass::runOnOperationAsync() PassUtils.cpp:0:0
#14 0x00007fec9742cc61 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) Pass.cpp:0:0
#15 0x00007fec9742d979 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) Pass.cpp:0:0
#16 0x00007fec9742f238 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) Pass.cpp:0:0
#17 0x00007fec9742d1a5 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) Pass.cpp:0:0
#18 0x00007fec9742d979 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) Pass.cpp:0:0
#19 0x00007fec97431d0a llvm::LogicalResult llvm::function_ref<llvm::LogicalResult (mlir::OpPassManager&, mlir::Operation*)>::callback_fn<mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_0>(long, mlir::OpPassManager&, mlir::Operation*) Pass.cpp:0:0
#20 0x00007fec994c314f mlir::iree_compiler::IREE::HAL::(anonymous namespace)::TranslateTargetExecutableVariantsPass::runOnOperation() TranslateExecutables.cpp:0:0
#21 0x00007fec9742cc61 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) Pass.cpp:0:0
#22 0x00007fec9742d979 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) Pass.cpp:0:0
#23 0x00007fec9742f238 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) Pass.cpp:0:0
#24 0x00007fec9742d1a5 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) Pass.cpp:0:0
#25 0x00007fec9742d979 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) Pass.cpp:0:0
#26 0x00007fec97431d0a llvm::LogicalResult llvm::function_ref<llvm::LogicalResult (mlir::OpPassManager&, mlir::Operation*)>::callback_fn<mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_0>(long, mlir::OpPassManager&, mlir::Operation*) Pass.cpp:0:0
#27 0x00007fec994c43e9 mlir::iree_compiler::IREE::HAL::(anonymous namespace)::TranslateAllExecutablesPass::runOnOperation() TranslateExecutables.cpp:0:0
#28 0x00007fec9742cc61 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) Pass.cpp:0:0
#29 0x00007fec9742d979 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) Pass.cpp:0:0
```

Environment: `iree-compile 3.12.0rc20260408` (d6ace956a5), gfx942

### PR DESCRIPTION
**Pull request by IanWood1**
Reject fusions that would broadcast away an entire attention dim group.

Closes https://github.com/iree-org/iree/issues/24051

### CODE REVIEW

**Review by MaheshRavishankar — APPROVED** [review]
This is a really neat solution! I was a bit apprehensive about "generality" of this, but actually there is a general solution here. Every op that implements the `IndexingMapOpInterface` (or `LinalgExt::FusionOpInterface` can provide a list of `importantDimGroups` and we can do this for any operation fusion.
